src/libstd/ffi/c_str.rs

```diff
@@ -632,6 +635,77 @@ impl CString {
         let this = mem::ManuallyDrop::new(self);
         unsafe { ptr::read(&this.inner) }
     }
+
+    /// Converts a `Vec` of `u8` to a `CString` without checking the invariants
+    /// on the given `Vec`.
+    ///
+    /// # Safety
+    ///
+    /// The given `Vec` **must** have one nul byte as its last element.
+    /// This means it cannot be empty nor have any other nul byte anywhere else.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use std::ffi::CString;
+    /// assert_eq!(
+    ///     unsafe { CString::from_vec_with_nul_unchecked(b"abc\0".to_vec()) },
+    ///     unsafe { CString::from_vec_unchecked(b"abc".to_vec()) }
+    /// );
+    /// ```
+    #[stable(feature = "cstring_from_vec_with_nul", since = "1.46.0")]
+    pub unsafe fn from_vec_with_nul_unchecked(v: Vec<u8>) -> Self {
+        Self { inner: v.into_boxed_slice() }
+    }
+
+    /// Attempts to converts a `Vec` of `u8` to a `CString`.
+    ///
+    /// Runtime checks are present to ensure there is only one nul byte in the
+    /// `Vec`, its last element.
+    ///
+    /// # Errors
+    ///
+    /// If a nul byte is present and not the last element or no nul bytes
+    /// is present, an error will be returned.
+    ///
+    /// # Examples
+    ///
+    /// A successful conversion will produce the same result as [`new`] when
+    /// called without the ending nul byte.
+    ///
+    /// ```
+    /// use std::ffi::CString;
+    /// assert_eq!(
+    ///     CString::from_vec_with_nul(b"abc\0".to_vec())
+    ///         .expect("CString::from_vec_with_nul failed"),
+    ///     CString::new(b"abc".to_vec())
+    /// );
+    /// ```
+    ///
+    /// A incorrectly formatted vector will produce an error.
+    ///
+    /// ```
+    /// use std::ffi::{CString, FromBytesWithNulError};
+    /// // Interior nul byte
+    /// let _: FromBytesWithNulError = CString::from_vec_with_nul(b"a\0bc".to_vec()).unwrap_err();
+    /// // No nul byte
+    /// let _: FromBytesWithNulError = CString::from_vec_with_nul(b"abc".to_vec()).unwrap_err();
+    /// ```
+    ///
+    /// [`new`]: #method.new
+    #[stable(feature = "cstring_from_vec_with_nul", since = "1.46.0")]
```

Comment:
unstable ?

Reply:
I didn't find rules about that, so I don't know what to pick. Is there somewhere which give heuristics for those things ?

Reply:
I think the rule is that if additions are small enough, they might be directly stabilized. However this seems like something the libs team might want to keep unstable for a bit, even if it is just because it needs bikeshedding for the name of the functions or things like that. I mean, even [`{f32, f64}::{LOG2_10, LOG10_2}`](https://github.com/rust-lang/rust/pull/69249) had to go through an unstable period 😄. But I'm sure of nothing, maybe this can be stabilized directly. @dtolnay or @cramertj can probably guide you for that.

What you will probably want to do is to mark these as unstable and [create a tracking issue](https://github.com/rust-lang/rust/issues/new?labels=C-tracking-issue&template=tracking_issue.md&title=Tracking+Issue+for+XXX). This will mean that the methods need a feature gate to be used and the libs team has freedom to make otherwise breaking changes to the methods.

I hope this helps :)

Reply:
Thanks @LeSeulArtichaut, I did just that !